[TensorExpr] Add a pass for trimming JIT graphs.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70475
* #70375
* #70374
* __->__ #66847
* #70535
* #66515
* #66514

Trimming means that we try to remove a small portion of the graph while
keeping it valid, and we try performing this step that N times. This is
useful for debugging when we try to find a minimal example reproducing
the issue at hand.

Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

Commit:
Add a pass for bisecting graphs.

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."


Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

Commit:
Update on "Add a pass for bisecting graphs."


Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "Add a pass for bisecting graphs."




Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "[TensorExpr] Add a pass for trimming JIT graphs."


Trimming means that we try to remove a small portion of the graph while
keeping it valid, and we try performing this step that N times. This is
useful for debugging when we try to find a minimal example reproducing
the issue at hand.

Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "[TensorExpr] Add a pass for trimming JIT graphs."


Trimming means that we try to remove a small portion of the graph while
keeping it valid, and we try performing this step that N times. This is
useful for debugging when we try to find a minimal example reproducing
the issue at hand.

Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "[TensorExpr] Add a pass for trimming JIT graphs."


Trimming means that we try to remove a small portion of the graph while
keeping it valid, and we try performing this step that N times. This is
useful for debugging when we try to find a minimal example reproducing
the issue at hand.

Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "[TensorExpr] Add a pass for trimming JIT graphs."


Trimming means that we try to remove a small portion of the graph while
keeping it valid, and we try performing this step that N times. This is
useful for debugging when we try to find a minimal example reproducing
the issue at hand.

Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

Commit:
Update on "[TensorExpr] Add a pass for trimming JIT graphs."


Trimming means that we try to remove a small portion of the graph while
keeping it valid, and we try performing this step that N times. This is
useful for debugging when we try to find a minimal example reproducing
the issue at hand.

Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

eellison commented:
ccc @davidberard98 -  something like this may be useful for repro'ing nvfuser failures

navahgar commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -916,18 +927,37 @@ BufHandle TensorExprKernel::bindSymbolicShapeInput(
+            inputTensorDims.emplace_back(
+                DimArg(bufHandle.dims()[i], "i" + c10::to_string(i)));
+          }
+          result = Compute(
```
Can we try to combine the copying here with the copying below? The one below includes handling strides, but that can probably be an argument and the logic of copying would be in one place then.

ZolotukhinM replied:
Let me see what we can do there!

ZolotukhinM replied:
Please take a look at #70535! I extracted the changes to a separate PR as you suggested below and refactored the logic a little.

navahgar commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -916,18 +927,37 @@ BufHandle TensorExprKernel::bindSymbolicShapeInput(
 
 Tensor TensorExprKernel::bindInput(const torch::jit::Value* input) {
   auto const& t = input->type();
+  auto const& outputs = input->owningGraph()->outputs();
```
Why are these changes to `kernel.cpp` in this PR? They are not directly related to "trimming JIT graphs", right?

ZolotukhinM replied:
That's correct, this fixes the behavior when a graph input is used directly as an output. We don't usually get such graphs, but that becomes common when we try trimming them.

I'll extract it to a separate PR.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/graph_opt.h:
```diff
@@ -69,6 +69,16 @@ TORCH_API std::shared_ptr<Graph> removeGraphOutput(
 
+// Perform \p ITERS rounds of "trimming" for the given \p GRAPH.
+//
+// Trimming means that we try to remove a small portion of the graph while
```
Tbh, that's exactly the contract this function provides. What exact portion is removed and how it is done is not guaranteed - that's an implementation detail and should not matter/be relied on.

Commit:
Update on "[TensorExpr] Add a pass for trimming JIT graphs."


Trimming means that we try to remove a small portion of the graph while
keeping it valid, and we try performing this step that N times. This is
useful for debugging when we try to find a minimal example reproducing
the issue at hand.

Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

navahgar reviewed (approved):
LGTM

Commit:
Update on "[TensorExpr] Add a pass for trimming JIT graphs."


Trimming means that we try to remove a small portion of the graph while
keeping it valid, and we try performing this step that N times. This is
useful for debugging when we try to find a minimal example reproducing
the issue at hand.

Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).
